base/iobuffer.jl

```diff
@@ -489,6 +489,75 @@ function take!(io::IOBuffer)
     return data
 end
 
+"Internal method. This method can be faster than takestring!, because it assumes
+the buffer is seekable, and because it does not reset the buffer to a usable state.
+Using the buffer after calling unsafe_takestring! may cause undefined behaviour.
+This function is meant to be used when the buffer is only used as a temporary
+string builder, which is discarded after the string is built."
+function unsafe_takestring!(io::IOBuffer)
+    off = io.offset
+    nbytes = io.size - off
+    iszero(nbytes) && return ""
+    # The C function can only copy from the start of the memory.
+    # Fortunately, in most cases, the offset will be zero.
+    return if iszero(off)
+        ccall(:jl_genericmemory_to_string, Ref{String}, (Any, Int), io.data, nbytes)
+    else
+        mem = StringMemory(nbytes)
+        unsafe_copyto!(mem, 1, io.data, off + 1, nbytes)
+        unsafe_takestring!(mem)
+    end
+end
+
+"""
+    takestring!(io::IOBuffer) -> String
+
+Return the content of `io` as a `String`, resetting the buffer to its initial
+state.
+This is preferred over calling `String(take!(io))` to create a string from
+an `IOBuffer`.
+
+# Examples
+```jldoctest
+julia> io = IOBuffer();
+
+julia> write(io, [0x61, 0x62, 0x63]);
+
+julia> s = takestring!(io)
+"abc"
+
+julia> isempty(take!(io)) # io is now empty
+true
+```
+
+!!! compat "Julia 1.12"
+    This function requires at least Julia 1.12.
+"""
+function takestring!(io::IOBuffer)
+    # If the buffer has been used up and needs to be replaced, there are no bytes, and
+    # we can return an empty string without interacting with the buffer at all.
+    io.reinit && return ""
+
+    # Currently, `IOBuffer`s (as opposed to `GenericIOBuffer`) is always seekable, since all
+    # IOBuffer constructors return seekable buffers, and GenericIOBuffer is internal.
+    # This makes `unsafe_takestring!` valid.
+    s = unsafe_takestring!(io)
+
+    # Restore the buffer to a usable state, making it no longer undefined behaviour to
+    # use the buffer after the `unsafe_takestring!` call.
+    # Note that if the buffer is not writable, there is no need to reinitialize the buffer,
+    # since it doesn't matter that the returned string looks into the same memory -
+    # this is because the buffer is not mutable through either the string nor the buffer.
+    if io.writable
+        io.reinit = true
+        io.mark = -1
+        io.ptr = 1
+        io.size = 0
+        io.offset = 0
+    end
+    s
+end
+
```

Comment:
Seems like there should also be a fallback method like:
```jl
takestring!(io::GenericIOBuffer) = String(take!(io))
```

Reply:
Good idea. Addressed in last commit.